test/test_common/simulated_time_system.cc

```diff
@@ -0,0 +1,194 @@
+    } else if (a->time() == b->time() && a->index() < b->index()) {
+      return true;
+    }
+  }
+  return false;
+};
+
+// Each timer is maintained and ordered by a common TimeSystem, but is
+// associated with a scheduler. The scheduler creates the timers with a libevent
+// context, so that the timer callbacks can be executed via Dispatcher::run() in
+// the expected thread.
+class SimulatedTimeSystem::SimulatedScheduler : public Scheduler {
+public:
+  SimulatedScheduler(SimulatedTimeSystem& time_system, Libevent::BasePtr& libevent)
+      : time_system_(time_system), libevent_(libevent) {}
+  TimerPtr createTimer(const TimerCb& cb) override {
+    return std::make_unique<SimulatedTimeSystem::Alarm>(time_system_, libevent_, cb);
+  };
+
+private:
+  SimulatedTimeSystem& time_system_;
+  Libevent::BasePtr& libevent_;
+};
+
+SimulatedTimeSystem::Alarm::~Alarm() {
+  if (armed_) {
+    disableTimer();
+  }
+}
+
+void SimulatedTimeSystem::Alarm::disableTimer() {
+  ASSERT(armed_);
+  time_system_.removeAlarm(this);
+  armed_ = false;
+}
+
+void SimulatedTimeSystem::Alarm::enableTimer(const std::chrono::milliseconds& duration) {
+  ASSERT(!armed_);
+  armed_ = true;
+  if (duration.count() == 0) {
+    activate();
+  } else {
+    time_system_.addAlarm(this, duration);
+  }
+}
+
+// When we initialize our simulated time, we'll start the current time based on
+// the real current time. But thereafter, real-time will not be used, and time
+// will march forward only by calling sleep().
+SimulatedTimeSystem::SimulatedTimeSystem()
+    : monotonic_time_(real_time_source_.monotonicTime()),
+      system_time_(real_time_source_.systemTime()), index_(0) {}
+
+SystemTime SimulatedTimeSystem::systemTime() {
+  Thread::LockGuard lock(mutex_);
+  return system_time_;
+}
+
+MonotonicTime SimulatedTimeSystem::monotonicTime() {
+  Thread::LockGuard lock(mutex_);
+  return monotonic_time_;
+}
+
+int64_t SimulatedTimeSystem::nextIndex() {
+  Thread::LockGuard lock(mutex_);
+  return index_++;
+}
+
+void SimulatedTimeSystem::addAlarm(Alarm* alarm, const std::chrono::milliseconds& duration) {
+  Thread::LockGuard lock(mutex_);
+  alarm->setTime(monotonic_time_ + duration);
+  alarms_.insert(alarm);
+}
+
+void SimulatedTimeSystem::removeAlarm(Alarm* alarm) {
+  Thread::LockGuard lock(mutex_);
+  alarms_.erase(alarm);
+}
+
+SchedulerPtr SimulatedTimeSystem::createScheduler(Libevent::BasePtr& libevent) {
+  return std::make_unique<SimulatedScheduler>(*this, libevent);
+}
+
+void SimulatedTimeSystem::setMonotonicTimeAndUnlock(const MonotonicTime& monotonic_time) {
+  // We don't have a convenient LockGuard construct that allows temporarily
+  // dropping the lock to run a callback. The main issue here is that we must
+  // be careful not to be holding mutex_ when an exception can be thrown.
+  // That can only happen here in alarm->activate(), which is run with the mutex
+  // released.
+  if (monotonic_time >= monotonic_time_) {
+    while (!alarms_.empty()) {
```

Comment:
nit: not sure if it is worth it or not. But I think this might be confusing to a new pair of eyes. It involves understanding that sets iterate in sorted order (and thanks to your overloaded operator this works), that we are modifying the set as we go, and that we are also moving time. I wonder if a comment spelling this out a bit would be helpful for others.

Reply:
Added comment
```
    // Alarms is a std::set ordered by wakeup time, so pulling off begin() each
    // iteration gives you wakeup order. Also note that alarms may be added
    // or removed during the call to activate() so it would not be correct to
    // range-iterate over the set.
```